packages/backend-common/src/reading/tree/ZipArchiveResponse.ts

```diff
@@ -0,0 +1,157 @@
+    private readonly stream: Readable,
+    private readonly subPath: string,
+    private readonly workDir: string,
+    private readonly filter?: (path: string) => boolean,
+  ) {
+    if (subPath) {
+      if (!subPath.endsWith('/')) {
+        this.subPath += '/';
+      }
+      if (subPath.startsWith('/')) {
+        throw new TypeError(
+          `ZipArchiveResponse subPath must not start with a /, got '${subPath}'`,
+        );
+      }
+    }
+  }
+
+  // Make sure the input stream is only read once
+  private onlyOnce() {
+    if (this.read) {
+      throw new Error('Response has already been read');
+    }
+    this.read = true;
+  }
+
+  private getPath(entry: Entry): string {
+    return entry.path.slice(this.subPath.length);
+  }
+
+  private shouldBeIncluded(entry: Entry): boolean {
+    if (this.subPath) {
+      if (!entry.path.startsWith(this.subPath)) {
+        return false;
+      }
+    }
+    if (this.filter) {
+      return this.filter(this.getPath(entry));
+    }
+    return true;
+  }
+
+  async files(): Promise<ReadTreeResponseFile[]> {
+    this.onlyOnce();
+
+    const files = Array<ReadTreeResponseFile>();
+
+    await this.stream
+      .pipe(unzipper.Parse())
+      .on('entry', (entry: Entry) => {
+        if (entry.type === 'Directory') {
+          entry.resume();
+          return;
+        }
+
+        if (this.shouldBeIncluded(entry)) {
+          files.push({
+            path: this.getPath(entry),
+            content: () => entry.buffer(),
+          });
+        } else {
+          entry.autodrain();
+        }
+      })
+      .promise();
+
+    return files;
+  }
+
+  async archive(): Promise<Readable> {
+    this.onlyOnce();
+
+    if (!this.subPath) {
+      return this.stream;
+    }
+
+    const archive = archiver('zip');
+    await this.stream
+      .pipe(unzipper.Parse())
+      .on('entry', (entry: Entry) => {
+        if (entry.type === 'File' && this.shouldBeIncluded(entry)) {
+          archive.append(entry, { name: this.getPath(entry) });
+        } else {
+          entry.autodrain();
+        }
+      })
+      .promise();
+    archive.finalize();
+
+    return archive;
+  }
+
+  async dir(options?: ReadTreeResponseDirOptions): Promise<string> {
+    this.onlyOnce();
+
+    const dir =
+      options?.targetDir ??
+      (await fs.mkdtemp(path.join(this.workDir, 'backstage-')));
+
+    await this.stream
+      .pipe(unzipper.Parse())
+      .on('entry', (entry: Entry) => {
+        if (this.shouldBeIncluded(entry)) {
+          const entryPath = this.getPath(entry);
+          if (entry.type === 'Directory') {
+            // Ignore directory entries since we handle that with the file entries
+            // since a zip can have files with directories without directory entries
+            entry.resume();
+            return;
+          }
+          const dirname = path.dirname(entryPath);
+          if (dirname) {
+            fs.mkdirSync(path.join(dir, dirname), { recursive: true });
```

Comment:
No sync please, this stops the entire backend all other plugins from any kind of execution until we've create this dir :>

Reply:
@Rugvip I tried to remove that now. I'm not sure that it will/should work though but the tests work. I would like some feedback on that and if it doesn't work I would appreciate some pointers on how to solve it.